File: organization-service/src/main/java/pl/piomin/services/organization/controller/OrganizationController.java
package pl.piomin.services.organization.controller;

import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Post;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import pl.piomin.services.organization.client.DepartmentClient;
import pl.piomin.services.organization.client.EmployeeClient;
import pl.piomin.services.organization.model.Organization;
import pl.piomin.services.organization.repository.OrganizationRepository;

import java.util.List;

@Controller("/organizations")
public class OrganizationController {

	private static final Logger LOGGER = LoggerFactory.getLogger(OrganizationController.class);
	
	OrganizationRepository repository;
	DepartmentClient departmentClient;
	EmployeeClient employeeClient;

	public OrganizationController(OrganizationRepository repository, DepartmentClient departmentClient, EmployeeClient employeeClient) {
		this.repository = repository;
		this.departmentClient = departmentClient;
		this.employeeClient = employeeClient;
	}

	@Post
	public Organization add(@Body Organization organization) {
		LOGGER.info("Organization add: {}", organization);
		return repository.add(organization);
	}
	
	@Get
	public List<Organization> findAll() {
		LOGGER.info("Organization find");
		return repository.findAll();
	}
	
	@Get("/{id}")
	public Organization findById(Long id) {
		LOGGER.info("Organization find: id={}", id);
		return repository.findById(id);
	}

	@Get("/{id}/with-departments")
	public Organization findByIdWithDepartments(Long id) {
		LOGGER.info("Organization find with departments: id={}", id);
		Organization organization = repository.findById(id);
		organization.setDepartments(departmentClient.findByOrganization(organization.getId()));
		return organization;
	}
	
	@Get("/{id}/with-departments-and-employees")
	public Organization findByIdWithDepartmentsAndEmployees(Long id) {
		LOGGER.info("Organization find with departments and employees: id={}", id);
		Organization organization = repository.findById(id);
		organization.setDepartments(departmentClient.findByOrganizationWithEmployees(organization.getId()));
		return organization;
	}
	
	@Get("/{id}/with-employees")
	public Organization findByIdWithEmployees(Long id) {
		LOGGER.info("Organization find with employees: id={}", id);
		Organization organization = repository.findById(id);
		organization.setEmployees(employeeClient.findByOrganization(organization.getId()));
		return organization;
	}
	
}

File: organization-service/src/main/kotlin/pl/piomin/services/organization/controller/OrganizationController.kt
package pl.piomin.services.organization.controller

import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.Post
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import pl.piomin.services.organization.client.DepartmentClient
import pl.piomin.services.organization.client.EmployeeClient
import pl.piomin.services.organization.model.Organization
import pl.piomin.services.organization.repository.OrganizationRepository

import javax.inject.Inject

@Controller("/organizations")
class OrganizationController(private val logger: Logger = LoggerFactory.getLogger(OrganizationController::class.java)) {

    @Inject
    lateinit var repository: OrganizationRepository
    @Inject
    lateinit var departmentClient: DepartmentClient
    @Inject
    lateinit var employeeClient: EmployeeClient

    @Post
    fun add(@Body organization: Organization): Organization {
        logger.info("Organization add: {}", organization)
        return repository.add(organization)
    }

    @Get
    fun findAll(): List<Organization> {
        logger.info("Organization find")
        return repository.findAll()
    }

    @Get("/{id}")
    fun findById(id: Long): Organization? {
        logger.info("Organization find: id={}", id)
        return repository.findById(id)
    }

    @Get("/{id}/with-departments")
    fun findByIdWithDepartments(id: Long): Organization? {
        logger.info("Organization find: id={}", id)
        val organization = repository.findById(id)
        if (organization != null)
            organization.departments = departmentClient.findByOrganization(organization.id)
        return organization
    }

    @Get("/{id}/with-departments-and-employees")
    fun findByIdWithDepartmentsAndEmployees(id: Long): Organization? {
        logger.info("Organization find: id={}", id)
        val organization = repository.findById(id)
        if (organization != null)
            organization.departments = departmentClient.findByOrganizationWithEmployees(organization.id)
        return organization
    }

    @Get("/{id}/with-employees")
    fun findByIdWithEmployees(id: Long): Organization? {
        logger.info("Organization find: id={}", id)
        val organization = repository.findById(id)
        if (organization != null)
            organization.employees = employeeClient.findByOrganization(organization.id)
        return organization
    }


}
